massa-execution-worker/src/interface_impl.rs

```diff
@@ -1330,6 +1332,167 @@ impl Interface for InterfaceImpl {
+    ///
+    /// # Returns
+    /// The id of the call
+    fn deferred_call_register(
+        &self,
+        target_addr: &str,
+        target_func: &str,
+        target_slot: (u64, u8),
+        max_gas: u64,
+        coins: u64,
+        params: &[u8],
+    ) -> Result<Vec<u8>> {
+        // This function spends coins + deferred_call_quote(target_slot, max_gas).unwrap() from the caller, fails if the balance is insufficient or if the quote would return None.
+
+        //    let total_booked_gas = get_current_total_booked_async_gas();
+
+        //    const CONST_ASYNC_GAS: u64 = XXXXX; // TODO calibrate: const_gas is the gas used even when gas=0 in order to process the item
+        //    let effective_gas = CONST_ASYNC_GAS + max_gas;
+        //    let fee = get_price(target_slot, effective_gas)?;
+
+        //    // make sender pay `coins + fee`
+
+        //    let call = /* ... */;
+        //    let id = /* ... */ ;
+
+        //    target_slot.booked_calls.push(callID, call);
+        //    target_slot.async_gas_booked += effective_gas;
+
+        //    set_current_total_booked_async_gas(total_booked_gas + effective_gas)
+
+        const CONST_DEFERRED_CALL_GAS: u64 = 0; // TODO calibrate: const_gas is the gas used even when gas=0 in order to process the item
+
+        let max_gas = CONST_DEFERRED_CALL_GAS + max_gas;
+
+        let target_addr = Address::from_str(target_addr)?;
+
+        // check that the target address is an SC address
+        if !matches!(target_addr, Address::SC(..)) {
+            bail!("target address is not a smart contract address")
+        }
+
+        // Length verifications
+        if target_func.len() > self.config.max_function_length as usize {
+            bail!("Function name is too large");
+        }
+        if params.len() > self.config.max_parameter_length as usize {
+            bail!("Parameter size is too large");
+        }
+
+        // check fee, slot, gas
+        let (available, fee_raw) = self.deferred_call_quote(target_slot, max_gas)?;
+        if !available {
+            bail!("The ASC call cannot be registered. Ensure that the target slot is not before/at the current slot nor too far in the future, and that it has at least max_gas available gas.");
+        }
+        let fee = Amount::from_raw(fee_raw);
+        let coins = Amount::from_raw(coins);
+
+        // write-lock context
+        let mut context = context_guard!(self);
+
+        // get caller address
+        let sender_address = match context.stack.last() {
+            Some(addr) => addr.address,
+            _ => bail!("failed to read call stack sender address"),
+        };
+
+        // make sender pay coins + fee
+        // coins + cost for booking the deferred call
+        context.transfer_coins(Some(sender_address), None, coins.saturating_add(fee), true)?;
+
+        let call = DeferredCall::new(
+            sender_address,
+            Slot::new(target_slot.0, target_slot.1),
```

Comment:
that was checked at line 1427 in `deferred_call_quote`

Reply:
Hm, I don't see the check in `deferred_call_quote` I also just see ```Slot::new(target_slot.0, target_slot.1),``` without checking the value of the thread?

Reply:
deferred_call_quote call `compute_call_fee` in `speculative_deferred_calls` 

```rust

        // Check that the slot is not in the past
        if target_slot <= current_slot {
            return Err(ExecutionError::DeferredCallsError(
                "Target slot is in the past.".into(),
            ));
        }

        // Check that the slot is not in the future
        if target_slot
            .slots_since(&current_slot, thread_count)
            .unwrap_or(u64::MAX)
            > async_call_max_booking_slots
        {
            // note: the current slot is not counted
            return Err(ExecutionError::DeferredCallsError(
                "Target slot is too far in the future.".into(),
            ));
        }

```

Reply:
I see. It's a bit hard to read for me, IMO it would be less error prone to:
- make the Slot::new() as soon as possible, and check that it is valid (thread under thread_count)
- only use this slot afterwards, and no (u64, u8)